### PR TITLE
[EFM] Robust error handling for Fallback State Machine

### PR DESCRIPTION
**Pull request by durkmurder**
https://github.com/onflow/flow-go/issues/6018

### Context

This PR updates error handling in `FallbackStateMachine` to handle all possible errors internally so a misbehaving smart contract cannot halt the chain by driving consensus participants into crash loop.

As part of this PR I have introduced a pubsub mechanism for orthogonal state machines, please consider it as initial implementation, I am very much open to suggestions on how we can improve the proposed design.

### CODE REVIEW

File: state/protocol/badger/state.go [hunk 1]
```diff
@@ -41,7 +41,7 @@ type State struct {
 	params                      protocol.Params
 	protocolKVStoreSnapshotsDB  storage.ProtocolKVStore
 	epochProtocolStateEntriesDB storage.EpochProtocolStateEntries // TODO remove when EpochProtocolStateEntry is stored in KVStore
-	protocolState               protocol.MutableProtocolState
+	protocolState               protocol.ProtocolState
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/badger/state.go:44]: Thanks for the context - very helpful. Also from my perspective, the current approach is totally fine. Essentially, `FollowerState` and `MutableProtocolState` just encapsulate a bunch of lower-level storage abstractions and implement higher-level logic on top, which is cleanly separated by our current design.

[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/badger/state.go:44]: A minor design oddity is the following cast from the interface to `badgerState.State` in all places where we instantiate a `FollowerState`:
https://github.com/onflow/flow-go/blob/557113c56be2deba08be5dbde260268a09275432/cmd/collection/main.go#L234-L247
I feel this is largely due to the fact that we use `badgerState.State` as a container for all the lower-level storage abstractions
*  Your code also uses it like that:  https://github.com/onflow/flow-go/blob/557113c56be2deba08be5dbde260268a09275432/state/protocol/badger/mutator.go#L79-L88
* I skimmed the code and it seems that the only place in `mutator.go` where we actually directly access a field in `badgerState.State` that is not a storage abstraction https://github.com/onflow/flow-go/blob/557113c56be2deba08be5dbde260268a09275432/state/protocol/badger/mutator.go#L792-L795 and this one could probably be encapsulated via a more refined interface (also improving clarity and modularity of the code)

I would be inclined to leave the code as is -- its only a minor aspect that could be improved in terms of clean abstractions. Nevertheless, just wanted to summarize my thoughts on how we could fix it if we wanted:
* Generally, the storage abstractions are initialized very early when booting a node. The storage abstractions (some subset of them) are injected into many many higher-level components. 
* If we had a container struct, that would bundle all the storage abstractions, we could:
  * more compactly represent the inputs for constructor methods: we input the container struct as opposed to each storage abstraction dependency separately (shorter list of method arguments for the constructors). The constructor just picks whatever they need
  * We don't need the concrete type (e.g. `badgerState.State`) just because we want to access the storage abstractions, as we do in the `FollowerState`.

File: state/protocol/protocol_state/epochs/fallback_statemachine_test.go [hunk 2]
```diff
@@ -663,3 +674,66 @@ func (s *EpochFallbackStateMachineSuite) TestEpochFallbackStateMachineInjectsMul
 			"final view should be greater than final view of test")
 	}
 }
+
+// TestProcessingMultipleEventsAtSameBlock tests that the state machine can process multiple events at the same block.
+// EpochFallbackStateMachineSuite has to be able to process any combination of events in any order at the same block.
+// This test generates a random amount of setup, commit and recover events and processes them in random order.
+func (s *EpochFallbackStateMachineSuite) TestProcesingMultipleEventsAtTheSameBlock() {
```
Comment (AlexHentschel) [inline state/protocol/protocol_state/epochs/fallback_statemachine_test.go:681]: could you please add ejection events here too? I think we have uncovered edge cases, where we should reject a recovery event that would be re-emitting an ejected node. Though, I think the current implementation would accept this edge case. 

You could add this as a ToDo to  https://github.com/onflow/flow-go/issues/6019. It would be more or less a test verifying that the current implementation has problems. My gut feeling is that the use if `partentState` causes the problems (not 100 sure that there is really an uncovered edge case through)

File: state/protocol/protocol_state/epochs/happy_path_statemachine.go [hunk 3]
```diff
@@ -58,12 +60,16 @@ func NewHappyPathStateMachine(view uint64, parentState *flow.RichEpochProtocolSt
 //     CAUTION: the HappyPathStateMachine is left with a potentially dysfunctional state when this error occurs. Do NOT call the Build method
 //     after such error and discard the HappyPathStateMachine!
 func (u *HappyPathStateMachine) ProcessEpochSetup(epochSetup *flow.EpochSetup) (bool, error) {
+	u.consumer.OnServiceEventReceived(epochSetup.ServiceEvent())
 	err := protocol.IsValidExtendingEpochSetup(epochSetup, u.parentState)
 	if err != nil {
```
Comment (AlexHentschel) [inline state/protocol/protocol_state/epochs/happy_path_statemachine.go:65]: similarly here. I think the error documentation is not strict enough yet to assume that _all_ errors are `InvalidServiceEventError` here (even though I think that is the case for the implementation -- but unless clearly documented, that is a property easily broken by accident)

File: state/protocol/protocol_state/epochs/happy_path_statemachine.go [hunk 4]
```diff
@@ -98,6 +104,7 @@ func (u *HappyPathStateMachine) ProcessEpochSetup(epochSetup *flow.EpochSetup) (
 	activeIdentitiesLookup := u.parentState.CurrentEpoch.ActiveIdentities.Lookup() // lookup NodeID → DynamicIdentityEntry for nodes _active_ in the current epoch
 	nextEpochActiveIdentities, err := buildNextEpochActiveParticipants(activeIdentitiesLookup, u.parentState.CurrentEpochSetup, epochSetup)
 	if err != nil {
```
Comment (AlexHentschel) [inline state/protocol/protocol_state/epochs/happy_path_statemachine.go:106]: similarly here: still ambiguity in the error specification (are really only `InvalidServiceEventError` possible?)

File: state/protocol/protocol_state/epochs/happy_path_statemachine.go [hunk 5]
```diff
@@ -124,25 +132,35 @@ func (u *HappyPathStateMachine) ProcessEpochSetup(epochSetup *flow.EpochSetup) (
 //     CAUTION: the HappyPathStateMachine is left with a potentially dysfunctional state when this error occurs. Do NOT call the Build method
 //     after such error and discard the HappyPathStateMachine!
 func (u *HappyPathStateMachine) ProcessEpochCommit(epochCommit *flow.EpochCommit) (bool, error) {
+	u.consumer.OnServiceEventReceived(epochCommit.ServiceEvent())
 	if u.state.NextEpoch == nil {
-		return false, protocol.NewInvalidServiceEventErrorf("protocol state has been setup yet")
+		err := protocol.NewInvalidServiceEventErrorf("protocol state has been setup yet")
```
Comment (AlexHentschel) [inline state/protocol/protocol_state/epochs/happy_path_statemachine.go:137]:
```suggestion
		err := protocol.NewInvalidServiceEventErrorf("received EpochCommit without prior EpochSetup")
```

File: state/protocol/protocol_state/epochs/happy_path_statemachine.go [hunk 6]
```diff
@@ -124,25 +132,35 @@ func (u *HappyPathStateMachine) ProcessEpochSetup(epochSetup *flow.EpochSetup) (
...
+		u.consumer.OnInvalidServiceEvent(epochCommit.ServiceEvent(), err)
+		return false, err
 	}
 	if u.state.NextEpoch.CommitID != flow.ZeroID {
-		return false, protocol.NewInvalidServiceEventErrorf("protocol state has already a commit event")
+		err := protocol.NewInvalidServiceEventErrorf("protocol state has already a commit event")
```
Comment (AlexHentschel) [inline state/protocol/protocol_state/epochs/happy_path_statemachine.go:142]:
```suggestion
		err := protocol.NewInvalidServiceEventErrorf("repeated EpochCommit event for epoch %d", epochCommit.Counter)
```

File: state/protocol/protocol_state/epochs/happy_path_statemachine.go [hunk 7]
```diff
@@ -58,12 +60,16 @@ func NewHappyPathStateMachine(view uint64, parentState *flow.RichEpochProtocolSt
...
 func (u *HappyPathStateMachine) ProcessEpochSetup(epochSetup *flow.EpochSetup) (bool, error) {
+	u.consumer.OnServiceEventReceived(epochSetup.ServiceEvent())
 	err := protocol.IsValidExtendingEpochSetup(epochSetup, u.parentState)
 	if err != nil {
+		u.consumer.OnInvalidServiceEvent(epochSetup.ServiceEvent(), err)
 		return false, fmt.Errorf("invalid epoch setup event: %w", err)
 	}
 	if u.state.NextEpoch != nil {
-		return false, protocol.NewInvalidServiceEventErrorf("repeated setup for epoch %d", epochSetup.Counter)
+		err := protocol.NewInvalidServiceEventErrorf("repeated setup for epoch %d", epochSetup.Counter)
```
Comment (AlexHentschel) [inline state/protocol/protocol_state/epochs/happy_path_statemachine.go:70]:
```suggestion
		err := protocol.NewInvalidServiceEventErrorf("repeated EpochSetup event for epoch %d", epochSetup.Counter)
```

File: state/protocol/protocol_state/epochs/happy_path_statemachine.go [hunk 8]
```diff
@@ -124,25 +132,35 @@ func (u *HappyPathStateMachine) ProcessEpochSetup(epochSetup *flow.EpochSetup) (
...
 	if u.state.NextEpoch.CommitID != flow.ZeroID {
-		return false, protocol.NewInvalidServiceEventErrorf("protocol state has already a commit event")
+		err := protocol.NewInvalidServiceEventErrorf("protocol state has already a commit event")
+		u.consumer.OnInvalidServiceEvent(epochCommit.ServiceEvent(), err)
+		return false, err
 	}
 	err := protocol.IsValidExtendingEpochCommit(epochCommit, u.parentState.EpochProtocolStateEntry, u.parentState.NextEpochSetup)
 	if err != nil {
+		u.consumer.OnInvalidServiceEvent(epochCommit.ServiceEvent(), err)
 		return false, fmt.Errorf("invalid epoch commit event: %w", err)
```
Comment (AlexHentschel) [inline state/protocol/protocol_state/epochs/happy_path_statemachine.go:149]:
```suggestion
		return false, fmt.Errorf("invalid epoch commit event for epoch %d: %w", epochCommit.Counter, err)
```

File: state/protocol/protocol_state/epochs/happy_path_statemachine.go [hunk 9]
```diff
@@ -58,12 +60,16 @@ func NewHappyPathStateMachine(view uint64, parentState *flow.RichEpochProtocolSt
 //     CAUTION: the HappyPathStateMachine is left with a potentially dysfunctional state when this error occurs. Do NOT call the Build method
 //     after such error and discard the HappyPathStateMachine!
 func (u *HappyPathStateMachine) ProcessEpochSetup(epochSetup *flow.EpochSetup) (bool, error) {
+	u.consumer.OnServiceEventReceived(epochSetup.ServiceEvent())
 	err := protocol.IsValidExtendingEpochSetup(epochSetup, u.parentState)
 	if err != nil {
+		u.consumer.OnInvalidServiceEvent(epochSetup.ServiceEvent(), err)
 		return false, fmt.Errorf("invalid epoch setup event: %w", err)
```
Comment (AlexHentschel) [inline state/protocol/protocol_state/epochs/happy_path_statemachine.go:67]:
```suggestion
		return false, fmt.Errorf("invalid epoch setup event for epoch %d: %w", epochSetup.Counter, err)
```

File: state/protocol/protocol_state/epochs/happy_path_statemachine.go [hunk 10]
```diff
@@ -124,25 +132,35 @@ func (u *HappyPathStateMachine) ProcessEpochSetup(epochSetup *flow.EpochSetup) (
...
 	return true, nil
 }
 
 // ProcessEpochRecover returns the sentinel error `protocol.InvalidServiceEventError`, which
 // indicates that `EpochRecover` are not expected on the happy path of epoch lifecycle.
-func (u *HappyPathStateMachine) ProcessEpochRecover(*flow.EpochRecover) (bool, error) {
-	return false, protocol.NewInvalidServiceEventErrorf("epoch recover event received while on happy path")
+func (u *HappyPathStateMachine) ProcessEpochRecover(epochRecover *flow.EpochRecover) (bool, error) {
+	u.consumer.OnServiceEventReceived(epochRecover.ServiceEvent())
+	err := protocol.NewInvalidServiceEventErrorf("epoch recover event received while on happy path")
```
Comment (AlexHentschel) [inline state/protocol/protocol_state/epochs/happy_path_statemachine.go:161]:
```suggestion
	err := protocol.NewInvalidServiceEventErrorf("epoch recover event for epoch %d received while on happy path", epochRecover.EpochSetup.Counter)
```

**Review by AlexHentschel — APPROVED** [review]
Very nice, clean code. Mostly suggestions regarding documentation and minor error handling improvements. Suggested to add a few references/TODOs to https://github.com/onflow/flow-go/issues/6019, but not entirely sure they are really correct. We could just dig further as part of https://github.com/onflow/flow-go/issues/6019. From my opinion, this PR is good to be merged.

File: state/protocol/protocol_state/consumer.go [hunk 11]
```diff
@@ -0,0 +1,23 @@
+package protocol_state
+
+import (
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// StateMachineEventsConsumer consumers notifications produced by OrthogonalStoreStateMachine instances. Any state machine
+// that performs processing of service events should notify the consumer about the events it received, successfully processed or
+// detected as invalid.
```
Comment (jordanschalm) [inline state/protocol/protocol_state/consumer.go:9]:
```suggestion
// detected as invalid.
// Implementations must:
//   - be concurrency safe
//   - be non-blocking
//   - handle repetition of the same events (with some processing overhead).
```
Suggesting to copy API requirements from the HotStuff consumer

**Review by jordanschalm — APPROVED** [review]
The improvements to the fallback state machine and new test are great - nice work!

I added a few suggestions regarding the event consumer for service events in the epoch state machines.

File: state/protocol/protocol_state/epochs/happy_path_statemachine.go [hunk 12]
```diff
@@ -124,25 +132,35 @@ func (u *HappyPathStateMachine) ProcessEpochSetup(epochSetup *flow.EpochSetup) (
...
 	}
 	if u.state.NextEpoch.CommitID != flow.ZeroID {
-		return false, protocol.NewInvalidServiceEventErrorf("protocol state has already a commit event")
+		err := protocol.NewInvalidServiceEventErrorf("protocol state has already a commit event")
+		u.consumer.OnInvalidServiceEvent(epochCommit.ServiceEvent(), err)
+		return false, err
 	}
 	err := protocol.IsValidExtendingEpochCommit(epochCommit, u.parentState.EpochProtocolStateEntry, u.parentState.NextEpochSetup)
 	if err != nil {
+		u.consumer.OnInvalidServiceEvent(epochCommit.ServiceEvent(), err)
```
[the remark this answers is not in the document]
Comment (durkmurder) [reply state/protocol/protocol_state/epochs/happy_path_statemachine.go]: I was considering it but it's only possible for `Received` and maybe `Processed` event, because internal state machines might handle errors internally so we actually don't know what is a successful or an invalid event.

File: state/protocol/protocol_state/epochs/fallback_statemachine.go
Comment (AlexHentschel) [inline state/protocol/protocol_state/epochs/fallback_statemachine.go:1]: hey, sorry could you extend the comment in [line 36](https://github.com/onflow/flow-go/blob/9e3192dff23c0ffe3a5cb5533bcba5f3132fe8d0/state/protocol/protocol_state/epochs/fallback_statemachine.go#L36) to the following:
```golang
// CAUTION: this logic must be consistent with the `EpochProtocolStateEntry.EpochPhase()`, which 
// determines the epoch phase based on the configuration of the fields we set here! 
// Specifically, if and only if the next epoch is already committed as of the parent state, 
// we go through with that committed epoch. Otherwise, we have a tentative values of an epoch 
// not yet properly specified, which we have to clear out. 
```

File: state/protocol/protocol_state/epochs/fallback_statemachine_test.go (hunk 2 above)
Comment (durkmurder) [reply state/protocol/protocol_state/epochs/fallback_statemachine_test.go:685]: I have added some stuff in [`6e09f32` (#6079)](https://github.com/onflow/flow-go/pull/6079/commits/6e09f324a93d35e40b329e0f2d779f940cadc301) but it will require extra love when implementing 6019. But you are right, currently there is an edge case and epoch recover event is accepted even though it readmits an ejected entity. We actually end up in broken state where current epoch has an entity ejected and in next epoch it's not.

Comment (AlexHentschel) [reply state/protocol/protocol_state/epochs/fallback_statemachine_test.go:685]: 👍 added [todo](https://github.com/onflow/flow-go/issues/6019#issuecomment-2189663711) to [issue 6019](https://github.com/onflow/flow-go/issues/6019)

File: state/protocol/protocol_state/epochs/fallback_statemachine.go [hunk 13]
```diff
@@ -108,39 +132,69 @@ func (m *FallbackStateMachine) ProcessEpochCommit(_ *flow.EpochCommit) (bool, er
...
-	if err != nil {
-		return false, fmt.Errorf("invalid epoch recovery event(setup): %w", err)
-	}
-
-	err = protocol.IsValidEpochCommit(&epochRecover.EpochCommit, &epochRecover.EpochSetup)
+	m.consumer.OnServiceEventReceived(epochRecover.ServiceEvent())
+	err := m.ensureValidEpochRecover(epochRecover)
 	if err != nil {
-		return false, fmt.Errorf("invalid epoch recovery event(commit): %w", err)
+		m.consumer.OnInvalidServiceEvent(epochRecover.ServiceEvent(), err)
```
[the remark this answers is not in the document]
Comment (durkmurder) [reply state/protocol/protocol_state/epochs/fallback_statemachine.go]: I've double checked everything, all errors reported to `OnInvalidServiceEvent` are `InvalidServiceEventError`s. Added extra documentation to nested functions to represent that.
